### ISSUE TITLE
Bundling with Parcel/npm

### ISSUE THREAD

**Issue by doconix** [TranscryptOrg/Transcrypt#596]
I just committed a branch that makes transcrypt work with [Parcel Bundler](https://parceljs.org/). Parcel is a zero-config bundler that is gaining traction. The branch is at:

https://github.com/QQuick/Transcrypt/tree/parcel-bundler

The example run is at `transcrypt/parcel-plugin-python/`. See the [README.md](https://github.com/QQuick/Transcrypt/blob/parcel-bundler/parcel-plugin-python/README.md) file for installation instructions.

When the test server is run, the parcel bundler starts with `index.html` and `index.js`, which contains:

```
import {main} from "./some.py";
main();
```

The `.py` extension tells parcel to run the file through transcrypt.

<soapbox>

I've said this before, but IMO transcrypt must work seamlessly with bundlers for it to see wide use. Modern JS projects include many coding languages (JS, CoffeeScript, TypeScript, Rust). The role of the bundler is to transpile them all to regular JS and then bundle them together.

Transcrypt is fundamentally different from bundler-based transpilers because it acts both as a transpiler AND a bundler.  It doesn't just transpile single files, but it also includes everything those file import. For example, if `some.py` contains "import mylib", transcrypt compiles BOTH `some.py` and `mylib.py`. A single run of transcrypt usually results in a number of .js files, which are placed together in `__target__`.

Regular transpilers generally stick to single files. When asked to compile `some.py`, they compile ONLY that file and don't follow any import statements. The import statements continue to keep the extensions of the files they import (transcrypt switches everything to .js). It's the job of the *bundler* to follow the imports and call the right transpiler.

So while this branch makes transcrypt work with Parcel, it is a bit of hack because transcrypt isn't meant to work this way. It also breaks the normal transcrypt process because it keeps import file extensions (in revisit* methods). See the diff in `compile.py` for where this happens.

The recent change to ES6 and units is a HUGE step toward doing things the bundler way, but a few things remain:

1. Transcrypt needs to keep import extensions (e.g. not change them). Like this:

Python Source:
```
from calculator import addit
```

JS Output:
```
import {addit} from './calculator.py';
```

This way the bundler knows to run transcrypt again for `calculator.py`.

2. Since transcrypt will only transpile a single .js file (the __main__ one), there's no need to copy other files to the `__target__` directory. This means the `__runtime__`, imported local files, and built-in modules (`re`, `time`, etc.) DON'T need to exist as siblings. Instead, these should be put into the npm repository just like any other imports. In fact, since only a single file export is needed, we can do away with `__target__` entirely.

An example import from the built-ins:

Source:
```
import re
```

JS Output:
```
import * as __module_re__ from 'transcrypt/re/__init__.py';
```

Note that the JS import is not relative (./), so the bundler will look in `node_modules/transcrypt/re/__init__.py` rather than in the local directory. It will expect the `re` module to be in the npm repository. Importing the `__runtime__` should be the same way.

I know JdeH wants to keep the current transpiler+bundler abilities of transcrypt, and I think that's fine. But I'd really like to see transcrypt become a first-order transpiler that works with bundlers out of the box. Right now it's not quite set up right to be one, but I don't think it needs a lot of changes to get there.

</soapbox>

**Comment by JdeH** [comment]
There isn't a doubt in my mind that this should be in somehow!
Can not yet oversee what this means for sourcemaps etc.
Seems we'll need the __target__ directory for those or ...?
And I'd like to retain the possibility to run Transcrypt stand-alone in a multi-module situation.
But that's a matter of adding a switch to enable cooperation with bundlers.
Any ideas?

**Comment by doconix** [comment]
Awesome! I am encouraged by your response and questions. I also can appreciate the need for backwards compatibility.

There are straightforward answers to your questions, but I can't write the response right now. I'll answer more in depth tomorrow. Things like the sourcemaps are done by the other languages (usually streamed, I think). The namespacing has a good solution too. Look for my response tomorrow.

**Comment by doconix** [comment]
FYI, here's the current npm stats on the bundlers. `webpack` is still the goliath and is going strong. Their last release supposedly did a lot to alleviate the "config hell" it required. My impression is the crazy configs of webpack put off a lot of python devs -- the entire thing was pretty orthogonal to python culture (e.g. one way to do things, focus on simplicity and clarity, etc.). Webpack requires a lot of wrangling in my experience. I guess the most recent release fixes much of that, but my personal work with it showed it was still a pain (although please factor in my bias against the wild west of JS culture, despite me living in the actual wild west :).

Parcel was released in Dec, 2017, so it's about a year old. It's an up-and-coming bundler, but I doubt it could ever overtake webpack (which has significant momentum right now). It advertises zero config and is a reaction against the webpack mess. My experiences with it over the last couple weeks is it is MUCH simpler than webpack, but it's still not the "clear and simple" that python devs would want. I do think it's the closest thing to it at this point. My opinion (which could probably easily be swayed) is that Parcel is currently the bundler that python devs would be most comfortable with.

But hopefully if we can get a bundler like Parcel working seamlessly, it will prepare transcrypt to be easily integrated into the other bundlers as well. Parcel plugins are pretty easy to code, although I've had to get into the actual parcel source code A LOT. Like most things JS, the documentation is less than stellar.

So again, here's the stats on the bundlers. You can see in the picture that Parcel doesn't really register. However, it's large number of stars is encouraging.

![screen shot 2018-11-26 at 10 33 46 am](https://user-images.githubusercontent.com/5728529/49025278-49fd5780-f169-11e8-8c6f-9e62ca926ee6.png)

**Comment by doconix** [comment]
> How to proceed with sourcemaps and the location to store them?

Let me try to summarize how the TypeScript parcel plugin works. It's a good example of how all the transpiler plugins work. 

1. Parcel sees an import statement in a JS file: `import * as bar from './foo.ts'`. This triggers the plugin to go to work.
1. The TS plugin loads the source from `foo.ts` into a variable and does the transpile.
1. The plugin needs to return a JS object containing the following to Parcel:
    1. `transpiled.outputText` - A string containing the output JS
    1. `transpiled.sourceMapText` - A string containing the source map in JSON format. This is converted 
    1. In some plugins, the result might also include the actual AST tree using `Babel AST` objects (a widely-used library).

Note that the TS plugin doesn't have to use the filesystem at all and doesn't have to stream anything between processes because it's written in JS. It runs right in the Parcel process and returns the objects.

Transcrypt can't do this because it needs to shell a Python process out of the plugin. That's not hard from JS/Node, but we need to figure out a good way to transfer back 1) the transpiled JS and 2) the sourcemap JSON. 

Remember that we adjust transcrypt to work like the other transpilers, it is transpiling the imported file ONLY. Not other .py files that the file might import. So we really only have TWO things to send back, NOT two things * the number of files.

It seems to me that streaming the result back to the plugin is the right way to do it so we don't clash with files on the filesystem. The stream could be JSON that contains the source + sourcemap. Alternatively, we could use two files placed in the `__target__` directory like transcrypt currently does. The plugin is doing it this way right now -- it deletes the `__target__` directory after reading the two files.

There are only two issues I saw with transpiling a single file rather than the entire an entire import tree. I mentioned them in my original post but I'll say them again because they apply to this point:

1. The import statements in the output JS source need to keep the file extensions of imported files. See my next post for the issue of resolution.
1. The python statement `from bar import *` requires information from bar.py because the JS doesn't have this concept.  The JS concept of `import * as {name} from 'foo.js'` is instead related to  python's `import foo`.  That means that the line `from bar import *` has to be transpiled as `import { e1, e2, e3, ... } from './foo.js`. We have to name each of the exported items in the imported file one by one, and that means we have to parse the imported file. This is doable when the imported file is Python, but it's a problem if the imported file is typescript or coffeescript or javascript.

One easy solution is to simply disallow the use of `from bar import *`. It's generally considered to be bad python anyway. This probably isn't the right thing to do long term, but you could punt on this one for the short term. Feel free to flame me on this one because I feel dirty even suggesting it. :)

**Comment by doconix** [comment]
> I get your point about modules having to reside in the npm repository, but what exactly does that mean for e.g. the re module? It won't be named `__init__.py` on npm, would it, since that name would occur many times. What, in your view, would the e.g re module look like on npm?

Hopefully the following gets the idea across -- it will take experimenting with examples to get the "final" layout, so this isn't an actual proposal. I'll do the `re` module since others are just siblings. I'm also thinking clean slate without regard to backwards compatibility at this point (so it will need adapting before going forward):

```
node_modules/
    transcrypt/
        src/
            modules/
                re/
                    __init__.py :: `import translate`
                    translate.py
        lib/
            modules/
                re.js :: "import {translate} from './re/translate.py';"
                re/
                    translate.js
```

My reading of npm best practices says both the original language (for reference) and the transpiled JS (for actual use) get published to npm. The `package.json` file of the package tells npm that the source to use for imports is in `transcrypt/lib/modules`. In other words, the JS always gets used during bundling. Other languages like TypeScript can import the modules without having to install Python. I realize there's no real use case for importing transcrypt modules into TypeScript, but doing it this way keeps us in line with typical npm expectations.

Before uploading an update to npm, we run transcrypt on the `modules/src/` directory so the two are in sync.

**Comment by doconix** [comment]
> Having modules like re, math and iterator on npm is bound to lead to name conflicts. But I'd like to keep the original Python module names and import syntax. However, a Python statement like import math simply doesn't contain enough info to prevent a name conflict if all such modules are stored on npm in "worldwide namespace". Transcrypt could add a prefix, but this cannot be expected from Javacript modules importing packages written in Transcrypt.

Since transcrypt does the resolution of imports in .py files, we have full control over this. The only issue I see is when python standard modules shadow over npm modules. I'll discuss that at the end.

Here's an example of various cases:

Python source: (`pother` is a local sibling python file, `jother` is a local sibling js file, `re` is a Python/transcript standard module, `moment` is a js library in npm repo):

```
import pother
import jother
import re
import moment
```

As transcrypt parses through the file, it decides how to resolve each one. The resulting JS import line determines where the bundler finds each file.

1. `import pother`: transcrypt finds `pother.py` in the source directory. It generates `import * as pother from './pother.py'`. (Parcel will call transcrypt later to transpile pother.py)
1. `import jother`: transcrypt doesn't find `pother.py` but finds `pother.js` in the source directory. It generates `import * as jother from './jother.js'`.
1. `import re`: transcrypt doesn't find `re.py` or `re.js` in the source directory. So it goes to the npm repo, first looking in the transcrypt lib. It finds `node_modules/transcrypt/lib/modules/re.js`. It generates `import * as re from 'transcrypt/lib/modules/re.js'`. 
1. `import moment`: transcrypt doesn't find `re.py` or `re.js` in the source directory. It doesn't find `re.js` in the transcrypt node_modules repo. It finds `node_modules/moment/`, a normal npm module. It generates `import * as moment from 'moment'`.

JS output:

```
import * as pother from './pother.py';
import * as jother from './jother.js';
import * as re from 'transcrypt/lib/modules/re.js';
import * as moment from 'moment';
```

So looking at the above, I think everything works beautifully with the one possible exception of a npm module being shadowed. Looking on npm, there *is* a module named [re](https://www.npmjs.com/package/re). That package can never be imported from python source code because transcrypt will always resolve to the standard python `re.js` (as above) and never to `node_modules/re/`. 

I would argue that this is an edge case, so perhaps the solution is as simple as `__pragma__ (‘js’, …)`?

**Comment by doconix** [comment]
Feel free to PM me if you want to set up a F2F call to discuss any of this. Also be sure to check the diff on the actual transcrypt source code (master branch vs. parcel-bundler branch). The changes are pretty minimal but were just a proof of concept rather than finished code. I think that diff shows how close transcrypt is to making this happen, but I'm leaving the decisions to you on how/what/where to modify the transcrypt code if we go forward with this kind of thing.

**Comment by doconix** [comment]
FYI I'm implementing some of these things today (`parcel-bundler` branch, not in `master`).

**Comment by doconix** [comment]
Status report: The changes to transcrypt to allow direct plugin use are almost done. I'll probably finish it tomorrow.

**Comment by JdeH** [comment]
@doconix 

I'd like to evaluate the changes you made, but have trouble getting an overview since lots of whitespace changes are also marked as differences. I've tried the ?w=0 option in the diff view but it doesn't solve the problem. Would it be possible for you to have only relevant (non-whitespace) etc. changes on GitHub, enabling comparison to the master branch. I've tried an external file compare tool but if failed on the  job, ignoring relevant differences. Sorry I have to ask this, but since the work on Transcrypt really takes up lots of my time now I feel forced to streamline. Know that I greatly appreciate all the effort you put in!

@PythonLinks @doconix

It's agreed that Transcrypt has to stay as simple as possible. I must say I find it impossibly hard to get any kind of grip upon the explosion of half-compatible, half-contemplated approaches in the JavaScript world. The prime reason Python is becoming so popular is its simplicity, although part of that recently is getting lost in overstressing facilities like iterators, decorators, metaclasses, static typing etc.. I hope Python is spared from getting as complex as e.g.  C++, although I really like that language and have used it most of my working life. Especially with Guido no longer at the wheel I am worried that the "the more the merrier" approach takes over. To be honest it isn't the low threshold common sense scripting language anymore that I once bumped into.

From the docs of Transcrypt:

http://www.transcrypt.org/docs/html/what_why.html#what-is-transcrypt

> One of the main goals of Transcrypt is to keep things simple, preventing it from disolution into an ocean of feature creep and conflicting demands. At its core it will stay lean and mean.

Wanting, or rather needing, this simplicity for me was the reason to abandon JS development of browsr apps and start the  Transcrypt project.

Transcrypt is primarily meant for Pythonista's who want to make browser applications in Python while using JS libs. Still the possibility of tight integration with the JS world is valuable. In this JS world an ever changing diversity of bundlers seems to be a fact of life. I don't expect these to converge as many JS developers don't seem to be particularly interested in that kind of thing. While popularity may shift, I don't expect it to ever really stabilize unless the ECMA committee proclaims a standard.

To support use of Transcrypt in combination with those, I propose the following criteria:

- The current simple way of working, including use of the flat __target__ directoryholding everything including sourcemaps, should be preserved as default.
- Changes aimed at making Transcrypt fit in the bundler world should not complicate TS's sourcecode too much, as it will all have to be maintained.
- Changes in Transcrypt itself should not be tied to one particular bundler. They should comply with emerging ECMA standards (still very virtual in this area), as does the module mechanism.

At the same time it would be very nice if Pythonista's could write e.g. datascience/AI related stuff in Python that would be available as JS compatible libraries that can be drawn into JavaScript apps the way JS programmers are used to, so using popular bundlers. This would enable Pythonista's to enter the enormous JS and Web Client world.

I want to be very clear: my heart is with Python, not with JavaScript. But since it's a massive stream, Pythonista's should be able to tap into it, even if I hope their main focus will be making algorithmically interesting server applications, rather than fancy front-ends whose main purpose is to look good.

**Comment by doconix** [comment]
@JdeH Thanks for the feedback on this. I've been at this the last couple days, and I'm pretty sure it will be done by today sometime. I'm almost there, but as you know, bugs can be a long tail. 

The changes to transcrypt are actually minimal, and I've been able to keep full backwards compatability. The bundle way of doing things is just an option. 

Some of the ideas I proposed didn't work as expected, so don't waste too much time trying to go through it. I think the finished code will be clear.

The implementation is very much in line with your 3 criteria. Well, except one. :) I was able to implement a deeper __target__ structure rather than flat. I think it's more clean this way because the target directory matches the source directory. BUT, you da boss, and I can step back to flat if you want. But perhaps you can wait until you see the running code to make a final decision...

I'll post again when done.

**Comment by doconix** [comment]
@JdeH Status update: It's working as of about an hour ago but I still need to do tests and ensure things are working. It will be another day or two. I have to step away from it for a bit.

**Comment by doconix** [comment]
@JdeH I am now working in the [parcel-bundler2](https://github.com/QQuick/Transcrypt/tree/parcel-bundler2) branch. I restarted from scratch (and copied relevant code from the other branch) so I could be back to a pristine copy of transcrypt.  I'm still putting the finishing touches on it, but I want to put out a few questions to you:

1. Right now I have the code in the transcrypt/parcel/ folder. Do you want this code in the transcrypt codebase, or do you want me to put it on a separate project? The plugin gets posted to npm, so it really doesn't matter where the source lives. It is really just a matter of preference on whether you want to keep things together or split them up.

2. I'd like to write a page in docs/ on using the parcel bundler, but again I want to ask if that is ok. Separate place or integrated with transcrypt docs?

3. Despite me making lots of changes to transcrypt's code in various tries, the end result is almost no changes!  The one change is in the constructor of `Program.__init__`:

The plugin needs to know what .py files transcrypt found and what .js files transcrypt created. I'm putting these in the `.options` file you create in `Program.__init__`.  The need for this is so the plugin can autoreload during development (recreate bundle anytime a .py file is changed). This is the ONLY thing that I had to change in the codebase, and all of it is in `compiler.py`.

The file now uses json format (instead of pickle) so my JS plugin can read it, and the name is changed: `<main>.options` -> `<main>.transcrypt.json`. There's no real need for that name change; I changed it simply to be closer to its new purpose. My reading of the code says that all the `pickleableOptions` are ok types for json. 

Let me know if it's OK to expand the options file like this.

4. No changes to command line arguments.

5. The plugin is using the flat structure. No need for deep structure. I did have that code working if you ever want to explore it, but flat works fine for the plugin.

6. I tested the built-in modules (`__runtime__`, `re`, etc.) on npm, but it actually worked better to keep them within the `__target__` directory. So no links to npm repo in the plugin.

7. The plugin is not tied directly into transcrypt, so other plugins and other bundlers hopefully can use (improve) the strategy I used for webpack, etc. No favoritism to this plugin. After a break, I can probably adapt it for webpack pretty easily. The current webpack plugin does work, but it's not seamless (imo).

Let me know about this code living inside transcrypt's github vs. separate project. And let me know about the docs. Regardless of your decisions on those two, I hope we can reference the docs where people can find them. I want to make it as easy as possible for people to use transcrypt in bigger projects.

**Comment by doconix** [comment]
@JdeH Just a ping on this one.

**Comment by JdeH** [comment]
@doconix 

I am attempting to run the examples, but at the first example I get:

Cannot find module 'parcel-bundler/src/Asset'

and

Cannot find module 'parcel-bundler/src/Logger'

I am using npm 6.5.0, parel-bundler 1.11.0, parcel-plugin-python 1.0.18, transcrypt 3.7.12

Any idea how to proceed? I'd love to find out how it all fits together...

**Comment by doconix** [comment]
@JdeH Sorry I've been out of touch on this. Thanks for spending time on it. I don't know why those files would be missing in the parcel-bundler package.... 

Anyway, agreed that we should discuss offline. Perhaps a screen share would be useful so I could show you what it's doing, and we can walk through each part.

I'd also like to mirror the ideas directly into webpack so transcrypt can work with both parcel and webpack (and the two plugins are similar to develop/maintain).

**Comment by doconix** [comment]
@JdeH I just posted a new update to the `parcel-bundler2` branch. This does the following:

1. Moved the testing files for Parcel into the `manual_tests` directory. Made adjustments to code and docs as needed to match.
2. Created a `Webpack` loader that mirrors the way the Parcel plugin works. 

One cool thing about the webpack loader is it is programmed in Transcrypt rather than JS. I tried switching the Parcel plugin to Transcrypt, but that has to extend the Parcel Asset JS class, so I reverted back to JS for it.

I still need to do the testing files for the webpack loader.

**Comment by pauleveritt** [comment]
Just curious, is it safe to write a tutorial showing this Parcel-based approach?

**Comment by pauleveritt** [comment]
As followup, [here's a tweet](https://twitter.com/jviide/status/1181265343408414720) from @jviide (with a tweet before it) showing VDOM and rendering in Python, as well as browser via Transcrypt. For folks interested in Python and SSR, it's an important step.

**Comment by JennaSys** [comment]
> Logger.js
> emoji.js
> prettyError.js

Is there any plan to fix the missing file issue?  Other than that, this Parcel plugin has been working great.

**Comment by JennaSys** [comment]
Following up, there was a change in the Parcel library where the logging module was moved.  It can be fixed by changing the reference to the logging module import in asset.js as I mentioned [here](https://github.com/QQuick/Transcrypt/issues/694#issuecomment-736174787).

**Comment by JennaSys** [comment]
Now that Parcel v2 is out and the current Parcel plug-in for Transcrypt is incompatible with the new version, it might be worth revisiting this to create a parcel-plugin-transcrypt-v2 that deprecates the old one that is broken anyway.

**Comment by JennaSys** [comment]
I just published a [Python transformer for Parcel V2](https://www.npmjs.com/package/parcel-transformer-transcrypt) that uses Transcrypt.  I'm still working on the docs and it needs _way_ more testing in different environments than I've done so far, but I figured I'd put it out there early and get some feedback.

Feel free to [open up a GitHub issue](https://github.com/JennaSys/parcel-transformer-transcrypt/issues) if you run into any problems with it.
